### ISSUE TITLE
Allow Autocomplete and RawAutocomplete to distinguish onFocus and onTap events

### ISSUE THREAD

**Issue by ekuleshov** [flutter/flutter#140662]
### Use case

Currently both `Autocomplete` and `RawAutocomplete` widgets trigger the `optionsBuilder` callback when widget becomes focused and then a dropdown with the list of options is immediately being shown after that.

However the widget could get focus in a several different ways:

* tap on the widget (both on mobile or web/desktop)
* using keyboard and tabbing through the form fields (this is especially relevant on web/desktop)

For the latter, it is not always desirable to have options immediately shown. Especially with an async `optionsBuilder` that may make a network call.

### Proposal

It would be great to have an option in both `Autocomplete` and `RawAutocomplete` widgets to control onFocus vs onTap behaviour.

Additionally it would make sense to have `AutocompleteNextOptionIntent` and `AutocompletePreviousOptionIntent` to trigger the `optionsBuilder`, so we could have custom UI (e.g. a dropdown chevron) or custom action shortcuts (e.g. Ctrl-Down as in common Windows shortcut for dropdowns fields) to open the list of options.

**Comment by DessisInformatica** [comment]
AutoComplete should have the FocusNode property settable.
Currently, it changes the FocusNode of TextField with each screen update, which compromises control.
If you set the FocusNode in the TextField, the component stops working, so the suggestion is to create the property in AutoComplete.

**Comment by timmaffett** [comment]
Hi @ekuleshov , #146995 adds the ability to set the `focusNode` and `textEditingController`.   

I wanted to discuss your proposal for `AutocompleteNextOptionIntent` and `AutocompletePreviousOptionIntent` to trigger the `optionsBuilder`..  This should be possible with #146995 because you will be able to trigger the `optionsBuilder` at any time by triggering a `notifyListeners()` on the `textEditingController`  (because `RawAutocomplete` places a listener on the `textEditingController` and triggers the options builder sequence whenever it is called..)
Additionally, you should be able to control the options popping up by just returning no options in the `optionsBuilder` callback whenever you do not want the options to display (and vice versa).  (In this way you could use a `Shortcuts` widget to wrap your view widget and catch Ctrl-DownArrow to trigger the options opening/and closing).

Does that make sense ?
I would like to keep #146995 as simple as possible and still also address this issue as well.

**Comment by timmaffett** [comment]
Dropdown activation is already tied to the focusNode and textEditingController within the `RawAutocomplete` widget.  Adding the ability to control adds a lot to the possible things that can be done...

This is some of the relevant (existing) code from `RawAutocomplete`:
```dart
  bool get _canShowOptionsView => _focusNode.hasFocus && _selection == null && _options.isNotEmpty;

  void _updateOptionsViewVisibility() {
    if (_canShowOptionsView) {
      _optionsViewController.show();
    } else {
      _optionsViewController.hide();
    }
  }

  // Called when _textEditingController changes.
  Future<void> _onChangedField() async {
    final TextEditingValue value = _textEditingController.value;
    final Iterable<T> options = await widget.optionsBuilder(value);
    _options = options;
    _updateHighlight(_highlightedOptionIndex.value);
    final T? selection = _selection;
    if (selection != null && value.text != widget.displayStringForOption(selection)) {
      _selection = null;
    }

    // Make sure the options are no longer hidden if the content of the field
    // changes (ignore selection changes).
    if (value.text != _lastFieldText) {
      _lastFieldText = value.text;
      _updateOptionsViewVisibility();
    }
  }
```

Simply doing a textEditController.notifyListeners() will trigger the listeners (no *actual* change is required in the text of the textEditController).
You could trigger the dropdown from another button by doing this:
1) foucsNode.requestFocus();
2) textEditController.notifyListeners(),
    [this will trigger a optionsBuilder callback]
3) then in the optionsBuilder callback we return a non-empty list (non empty list will make the options dropdown appear, or an empty list will make the options dropdown disappear)

[Refer  `_canShowOptionsView` and `_onChangedField` in the code above to see how this will trigger showing/hiding the options dropdown]

You would place/wrap tap listeners on whichever widgets you wanted to be able to do this (for example on the dropdown chevron).  You could place shortcut listeners higher up in the page and capture Ctrl-DownArrow shortcuts and then see whats focused and do the above to trigger the dropdown...
You could also place listeners on focusNode so that you could control onFocus behaviors. 

Using this method of controlling the options dropdown visibility to triggering/and controlling the # of options in the options list should all complete control of when the options list appears/disappears.

I understand that being able to control the shortcut and intent list that `RawAutocomplete` create might be more versatile and ideal, but I don't think that level of change, to a lower level widget in the Flutter framework, would not be an easy change to get through the approval process.  (this change may even be difficult to get through and it is just passing 2 arguments through to `RawAutocomplete`).

I try and make some time to play with making an example proof of concept.

**Comment by timmaffett** [comment]
Here is a dartpad example of a proof of concept:

https://dartpad.dev/?id=f679a62ef5aba7d31e39569cf98e4432

**Comment by ekuleshov** [comment]
> Here is a dartpad example of a proof of concept:
> 
> https://dartpad.dev/?id=f679a62ef5aba7d31e39569cf98e4432
> 
> This is the text field in app header RawAutocomplete example with two things added:
> 
> 1. chevron to open/close the options dropdown.
> 2. clicking on the text fields automatically opens the dropdown.

Something off with that example. It does print something about focus changes, but I was unable to see the keyboard-based dropdown interaction I described above (esc key doesn't work, and the dropdown is not getting focus upon activation).

Though generally I see several issues with this approach:

* it is kind of backwards - instead of having a definitive action/intent to show dropdown you have to fiddle with focus and text controller to show the dropdown as a side effect of the option rebuild
* it is hard to tell in the option builder if rebuild was called due to some arbitrary focus change or if it was triggered by user
* you have to create another widget wrapper and fiddle with state management which you generally don't control. I.e. you have to make assumptions how focus and text controller being updated, and they are updated by the widget you are wrapping in
* it is based on non-public API

**Comment by timmaffett** [comment]
@ekuleshov  fair enough.  I was trying to illustrate that the addition of `focusNode` and `textEditingController` allowed more possibilities.  What you are trying to accomplish requires rewrites of `RawAutocomplete` (and `Autocomplete`) at a much bigger level - adding intents and shortcuts for example to handle the keyboard interaction.

I will concede that PR #146995 does not go far enough accomplish what you desire, and i definitely do not want to make it any more complex to do so.

**Comment by justinmc** [comment]
@victorsanni This is related to your work on controlling when the options view shows.
